feat(exo): add support for eXO systems

flz opened the pull request:
Let's start with this, I need to look at feedback received recently.

flz committed:
Add draft support for eXO

flz committed:
Add draft support for eXO

flz committed:
Add draft support for eXO

flz committed:
Merge branch 'master' into exo

flz committed:
More tests and lint fixes

flz committed:
More tests and lint fixes

flz committed:
More tests and lint fixes

ozrex75 committed:
Refresh token if expired

Token expires after 8 hours.  Refresh the token if the request receives an unauthorised response.

flz committed:
Merge branch 'master' into exo

flz committed:
Rename Toggle devices to Switches, rework logic

flz committed:
Rename Toggle devices to Switches, rework logic

flz committed:
Rename Toggle devices to Switches, rework logic

flz committed:
Update tests to use new classes

flz committed:
Update tests to use new classes

flz committed:
Update tests to use new classes

flz committed:
Some isort fixes

flz committed:
Some isort fixes

flz committed:
Some isort fixes

flz committed:
Now some mypy --strict fixes

flz committed:
Now some mypy --strict fixes

flz committed:
Now some mypy --strict fixes

flz committed:
Merge remote-tracking branch 'origin/exo' into exo

flz committed:
Merge branch 'master' into exo

flz committed:
Merge branch 'master' into exo

flz committed:
Merge remote-tracking branch 'origin/exo' into exo

TheKamakaZi commented:
Hey @flz,

Been using the eXO branch of your integration for the past 2 months without issue. Would it be possible to merge this to master so that it can be included in HA's core? 

Cheers, and thanks for all your hard work!

malatg commented:
Is there any plans you incoporate this into the mail branch?

flz commented:
Somehow I missed the comments here.

I'll rebase on top of the main branch. If things still work, I'll release a new version to make this available through HA.

flz committed:
Merge branch 'master' into exo

malatg commented:
Awesome! I hope it works 😊

flz commented:
> Awesome! I hope it works 😊

You let me know, I don't have exo systems :-)

malatg commented:
Hi,

I am sorry of I have misunderstood  something but when will this be included in the main branch?

/Marcus

flz commented:
@malatg I meant I can't test the exo support since I don't have that hardware. If you can test the exo branch and let me know it works, I'll merge it to master.

malatg commented:
I can definitely test it if you can point me in the direction on how I test the branch? I mean how do I switch the to this branch for that particular integration?

TheKamakaZi commented:
Hey @flz,

Thanks for this! I'll deploy tonight, run some tests in the morning, and let you know.

@malatg, simply switch to the `exo` branch after cloning, and restart HA.

danricho commented:
My only issue with this is that my pump `data={'name': 'filter_pump', 'type': 1, 'state': 0}`, will not change state by the `.turn_on()` method even if I force it to an `ExoAuxSwitch` or `ExoAttributeSwitch`.  

As the current implementation stands, it is defaulting to an `ExoAttributeSensor`.

flz commented:
@danricho Can you provide me with some more debug logging (the complete API output of systems/devices would be useful).

Veuchez commented:
Excuse my ignorance, but where should these files be copied into home assistant?

mk-french commented:
Hey @flz , is there a particular branch of your HA core fork to use with this branch?

I've been using my forks [iaqualink-py/test-wo-shadow-client](https://github.com/mk-french/iaqualink-py/tree/test-wo-shadow-client) and [core/iaqualink-eXO](https://github.com/mk-french/core/tree/iaqualink-eXO) without any real issues for a while now but happy to help test this and provide some feedback to get it merged - then migrate over.

There's a few errors when using this branch and your [core/iaqualink-http2](https://github.com/flz/home-assistant/tree/iaqualink-http2) component. At initial glance they're related to type conversions etc. I haven't yet sifted through them in detail to debug but wanted to make sure I was starting in the right place first... :)

-Martin

flz committed:
Merge branch 'master' into exo

flz committed:
Merge branch 'master' into exo

flz committed:
Lint fixes

flz commented:
> Hey @flz , is there a particular branch of your HA core fork to use with this branch?

I don't think there have been API changes in the iaqualink-py libraries for exo support but I'd need to double check. It's been a while since I looked at the code.

> I've been using my forks [iaqualink-py/test-wo-shadow-client](https://github.com/mk-french/iaqualink-py/tree/test-wo-shadow-client) and [core/iaqualink-eXO](https://github.com/mk-french/core/tree/iaqualink-eXO) without any real issues for a while now but happy to help test this and provide some feedback to get it merged - then migrate over.
> 
> There's a few errors when using this branch and your [core/iaqualink-http2](https://github.com/flz/home-assistant/tree/iaqualink-http2) component. At initial glance they're related to type conversions etc. I haven't yet sifted through them in detail to debug but wanted to make sure I was starting in the right place first... :)

I'll refresh the core/exo branch and make sure iaqua support works well. I'm at the point where I'm tempted to ship exo support as experimental in the library, just to make it easier for people to test.

flz commented:
I've updated the core/exo branch.

The two ways to test I can think of:

Separate test home-assistant instance
* Clone https://github.com/flz/home-assistant, switch to exo branch
* Open in VSCode, then do Reopen in Container
* Run python -m homeassistant in the VSCode terminal
* Access http://localhost:8123/ in your browser and go through regular set up, add iaqualink integration.

Modify your existing HA installation
* Update homeassistant/components/iaqualink/manifest.json and replace the requirement with "git+https://github.com/flz/iaqualink-py.git@exo#iaqualink==0.5.1".
* You may have to restart homeassistant for it to be picked up.

flz committed:
Merge branch 'master' into exo

flz committed:
Fix new lint issues

mk-french committed:
Fix Exo bugs

Intent is to maintain existing API compatability with Homeassistant.
- Fix bug referencing incorrect 'Sensor' for temperature PV.
- Add temperature unit
- Ignore serial number and versions (issues parsing strings into floats)
- Extract the heater state feedback into it's own 'device' to align with the iAqua heater control structure, to allow the state property to hold the temperature SP.
- Associated test fixes

flz committed:
Add test script to list systems and devices

flz committed:
Merge branch 'master' into exo

# Conflicts:
#	pyproject.toml

flz committed:
Merge branch 'master' into exo

# Conflicts:
#	pyproject.toml

flz committed:
Some lint fixes

flz committed:
Some more fixes for the test script

flz commented:
I've added a test script to see if the library is parsing the systems/devices properly. You can run it with `hatch run dev:test` if you have hatch installed (or install it with these [instructions](https://hatch.pypa.io/1.13/install/)). Otherwise, you can run it directly but you'll need to add the necessary dependencies.

I haven't been able to test the script with device enumeration since my iaqualink system is offline and I suspect it's dead).

flz committed:
Merge branch 'master' into exo

flz committed:
Merge branch 'master' into exo

# Conflicts:
#	pyproject.toml

flz commented:
Here's some information on how to test the current exo code: https://github.com/flz/iaqualink-py/wiki/eX0-development-branch-testing

flz committed:
Merge branch 'master' into exo

# Conflicts:
#	pyproject.toml

flz committed:
Syle fixes

flz committed:
Merge branch 'master' into exo

# Conflicts:
#	pyproject.toml

malatg commented:
Hi @Whytey could you give me the exact steps (in detail 😇) on how you achieved the above?

Thank you in advance!

Whytey commented:
I was running a custom version of the iaqualink integration.  I had my cholorinator already added to HA, though it had stopped working with my upgrade to 2025.6.  I am running HA in HAOS.

HA was running and I did __not__ delete the old integration from within HA before running the follow commands:
```
cd config
mkdir temp
mv custom_components/iaqualink temp
docker exec -it homeassistant bash
vi /usr/src/homeassistant/homeassistant/components/iaqualink/manifest.json 
```
Change the `iaqualink` requirement so it reads `"git+https://github.com/flz/iaqualink-py.git@exo#iaqualink==0.5.1".`

Restart HA and everything should be present.

Hope that helps,
Whytey

Hrvojeee commented:
Hi guys? Did anyone test so this could be merged?

flz committed:
Merge branch 'master' into exo

# Conflicts:
#	pyproject.toml
#	uv.lock

flz committed:
chore: remove commented test, add reason for skipped tests

flz commented:
A quick note that if you're currently using the exo support, please subscribe to this [discussion](https://github.com/flz/iaqualink-py/discussions/203) as I will be using it to ask for testing on occasion.
